spark/src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableReplaceTests.scala

```diff
@@ -0,0 +1,745 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.functions.{array, col, map, struct}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{ArrayType, MapType, StructType}
+
+trait DeltaAlterTableReplaceTests extends DeltaAlterTableTestBase {
+
+  import testImplicits._
+
+  ddlTest("REPLACE COLUMNS - add a comment") {
+    val df = Seq((1, "a"), (2, "b")).toDF("v1", "v2")
+      .withColumn("s", struct("v1", "v2"))
+      .withColumn("a", array("s"))
+      .withColumn("m", map(col("s"), col("s")))
+
+    withDeltaTable(df) { tableName =>
+
+      sql(s"""
+             |ALTER TABLE $tableName REPLACE COLUMNS (
+             |  v1 int COMMENT 'a comment for v1',
+             |  v2 string COMMENT 'a comment for v2',
+             |  s STRUCT<
+             |    v1:int COMMENT 'a comment for s.v1',
+             |    v2:string COMMENT 'a comment for s.v2'> COMMENT 'a comment for s',
+             |  a ARRAY<STRUCT<
+             |    v1:int COMMENT 'a comment for a.v1',
+             |    v2:string COMMENT 'a comment for a.v2'>> COMMENT 'a comment for a',
+             |  m MAP<STRUCT<
+             |      v1:int COMMENT 'a comment for m.key.v1',
+             |      v2:string COMMENT 'a comment for m.key.v2'>,
+             |    STRUCT<
+             |      v1:int COMMENT 'a comment for m.value.v1',
+             |      v2:string COMMENT 'a comment for m.value.v2'>> COMMENT 'a comment for m'
+             |)""".stripMargin)
+
+      val deltaLog = getDeltaLog(tableName)
+      val expectedSchema = new StructType()
+        .add("v1", "integer", true, "a comment for v1")
+        .add("v2", "string", true, "a comment for v2")
+        .add("s", new StructType()
+          .add("v1", "integer", true, "a comment for s.v1")
+          .add("v2", "string", true, "a comment for s.v2"), true, "a comment for s")
+        .add("a", ArrayType(new StructType()
+          .add("v1", "integer", true, "a comment for a.v1")
+          .add("v2", "string", true, "a comment for a.v2")), true, "a comment for a")
+        .add("m", MapType(
+          new StructType()
+            .add("v1", "integer", true, "a comment for m.key.v1")
+            .add("v2", "string", true, "a comment for m.key.v2"),
+          new StructType()
+            .add("v1", "integer", true, "a comment for m.value.v1")
+            .add("v2", "string", true, "a comment for m.value.v2")), true, "a comment for m")
+      assertEqual(deltaLog.snapshot.schema, expectedSchema)
+
+      implicit val ordering = Ordering.by[
+        (Int, String, (Int, String), Seq[(Int, String)], Map[(Int, String), (Int, String)]), Int] {
```

Comment:
Maybe worth a type alias? That tuple is a real mouthful to type twice!

(several more below)

Reply:
Can't really add a type alias, because the ordering and contents of the type keeps changing in all the tests :(